### DISCUSSION TITLE
Should there be a CreateOrUpdateStreamAsync

### DISCUSSION THREAD

**Discussion by alberk8** [nats-io/nats.net#466] [Q&A]
There is a CreateOrUpdateConsumerAsync, is it possible to have **CreateOrUpdateStreamAsync**  ?.

**Comment by mtmk** [comment]
They currently map to JetStream API call and AFAIK `STREAM.CREATE` does not allow for updating existing streams but it is idempotent and would return the stream if already created with the same configuration so I don't think you can update a stream with this call if you provided different config, can you? Whereas consumer API would allow some of the configuration to be changed.

**Reply by alberk8** [reply]
Would this be some sort of an equivalent?  

```
StreamConfig streamConfig = new StreamConfig("EVENTS", new[] { "data.>" })
 {    
 };

INatsJSStream stream;

var evnt = await jsContext.GetStreamAsync("EVENTS");

if (evnt != null)
{
    Debug.WriteLine("UpdateStreamAsync");
    streamConfig.MaxAge = TimeSpan.FromMinutes(30);
    stream = await jsContext.UpdateStreamAsync(streamConfig);
}
else
{
    Debug.WriteLine("CreateStreamAsync");
    stream = await jsContext.CreateStreamAsync(new StreamConfig("EVENTS", new[] { "data.>" }));
}
```

**Reply by mtmk** [reply]
It's a good idea which would make the API more uniform. Of course it's not the same as consumers i.e. consumer 'upsert' is probably done atomically etc. on the server side. @caleblloyd wdyt?

edit: some other points; it's an expensive call with stream info, it has race conditions. it should really be done on the server to be same as consumers. Maybe it belongs to a higher level api?

**Reply by alberk8** [reply]
What you are saying is that  changing the stream config can cause existing consumer of the stream elsewhere (on another service/app) to experience error?

**Reply by mtmk** [reply]
AFAIU getting the stream info quite an expensive operation for the server, collecting stats etc. that would be the main issue.

**Reply by caleblloyd** [reply]
I think you may be able to achieve upsert-like functionality with:

```csharp
INatsJSStream stream;
try
{
    stream = await js.UpdateStreamAsync(config);
}
catch (NatsJSApiException e) when (e.Error.Code is 404)
{
    stream = await js.CreateStreamAsync(config);
}
```

I have mixed feelings about putting this in the library as `CreateOrUpdateStreamAsync` though because as @mtmk points out this is not an atomic server-side operation.
